## AND-102 Task 5: Merge 1, License + Installed — Part 1

In [1]:
import pandas as pd

license_df = pd.read_csv("../data/license.csv")


KEEP = {"ACTIVE", "PENDING_RENEWAL"}

license_df = license_df[license_df["LICENSESTATUS"].isin(KEEP)].reset_index(drop=True)
print(license_df.shape)
license_df.head()

(43297, 11)


,ElevatingDevicesNumber,LocationoftheElevatingDevice,ElevatingDevicesLicenseNumber,LICENSESTATUS,LICENSEEXPIRYDATE,LICENSEHOLDER,LICENSEHOLDERACCOUNTNUMBER,LICENSEHOLDERADDRESS,BILLINGCUSTOMER,BILLINGADDRESS,BILLINGACCOUNT
0,10,111 WELLESLEY ST W TORONTO M7A 1A2 ON CA,EDLIC-000010,ACTIVE,28-Apr-17,LEGISLATIVE ASSEMBLY OF ONTARIO ATTN: JOHN ED...,data redacted,99 WELLESLEY ST W WHITNEY BLOCK ROOM 2540 TOR...,LEGISLATIVE ASSEMBLY OF ONTARIO ATTN: JOHN ED...,99 WELLESLEY ST W WHITNEY BLOCK ROOM 2540 TOR...,data redacted
1,1009,404 MAIN ST WOODSTOCK N4S 7X5 ON CA,EDLIC-001009,ACTIVE,15-Jul-17,AGRIBRANDS PURINA CANADA INC,data redacted,404 MAIN ST PO BOX 250 WOODSTOCK ON N4S 7X5 CA,AGRIBRANDS PURINA CANADA INC,"404 MAIN ST PO BOX 250 WOODSTOCK, ON, N4S 7X5, CA",data redacted
2,10145,45 SECOND ST E CORNWALL K6H 1V5 ON CA,EDLIC-010145,ACTIVE,15-Mar-17,CORNWALL PUBLIC LIBRARY,data redacted,45 SECOND ST E P.O.BOX 939 CORNWALL ON K6H 5V1 CA,CORNWALL PUBLIC LIBRARY,"45 SECOND ST E P.O.BOX 939 CORNWALL, ON, K6H 5...",data redacted
3,1018,150 SIMCOE ST LONDON N6A 4M3 ON CA,EDLIC-001018,ACTIVE,15-Oct-17,LABATT LONDON -LONDON BREWERY,data redacted,150 SIMCOE ST PO BOX 5050 STN B LONDON ON N6A ...,LABATT LONDON -LONDON BREWERY,"150 SIMCOE ST PO BOX 5050 STN B LONDON, ON, N6...",data redacted
4,1019,150 SIMCOE ST LONDON N6A 4M3 ON CA,EDLIC-001019,ACTIVE,15-Oct-17,LABATT LONDON -LONDON BREWERY,data redacted,150 SIMCOE ST PO BOX 5050 STN B LONDON ON N6A ...,LABATT LONDON -LONDON BREWERY,"150 SIMCOE ST PO BOX 5050 STN B LONDON, ON, N6...",data redacted


In [2]:
installed_df = pd.read_json("../data/installed.json")

print(installed_df.shape)
installed_df.head()

(46936, 9)


,Elevating devices number,Owner Name,Owner Address,Owner Account Number,Device Class,Device Type,DeviceStatus,Location of Device,under review
0,8,LEGISLATIVE ASSEMBLY OF ONTARIO ATTN: JOHN ED...,99 WELLESLEY ST W WHITNEY BLOCK ROOM 2540 TOR...,redacted,Elevators,Passenger Elevator,Active,111 WELLESLEY ST W TORONTO M7A 1A2 ON CA,N
1,9,LEGISLATIVE ASSEMBLY OF ONTARIO ATTN: JOHN ED...,99 WELLESLEY ST W WHITNEY BLOCK ROOM 2540 TOR...,redacted,Elevators,Passenger Elevator,Active,111 WELLESLEY ST W TORONTO M7A 1A2 ON CA,N
2,10,LEGISLATIVE ASSEMBLY OF ONTARIO ATTN: JOHN ED...,99 WELLESLEY ST W WHITNEY BLOCK ROOM 2540 TOR...,redacted,Elevators,Passenger Elevator,Active,111 WELLESLEY ST W TORONTO M7A 1A2 ON CA,N
3,11,LEGISLATIVE ASSEMBLY OF ONTARIO ATTN: JOHN ED...,99 WELLESLEY ST W WHITNEY BLOCK ROOM 2540 TOR...,redacted,Elevators,Passenger Elevator,Active,111 WELLESLEY ST W TORONTO M7A 1A2 ON CA,N
4,13,LEGISLATIVE ASSEMBLY OF ONTARIO ATTN: JOHN ED...,99 WELLESLEY ST W WHITNEY BLOCK ROOM 2540 TOR...,redacted,Elevators,Passenger Elevator,Active,111 WELLESLEY ST W TORONTO M7A 1A2 ON CA,N


In [3]:
# Merge 1 start — row counts before joining
print(f'license_df:   {len(license_df):,} rows')
print(f'installed_df: {len(installed_df):,} rows')


license_df:   43,297 rows
installed_df: 46,936 rows


## AND-102 Task 5: Merge 1, License + Installed — Part 2

In [4]:
# Common key: 'ElevatingDevicesNumber' (license) <-> 'Elevating devices number' (installed)
rows_license   = len(license_df)
rows_installed = len(installed_df)

merged_df = license_df.merge(
    installed_df,
    left_on='ElevatingDevicesNumber',
    right_on='Elevating devices number',
    how='inner'
)

print(f'license_df rows:   {rows_license:,}')
print(f'installed_df rows: {rows_installed:,}')
print(f'merged_df rows:    {len(merged_df):,}')

merged_df.head()

license_df rows:   43,297
installed_df rows: 46,936
merged_df rows:    43,297


,ElevatingDevicesNumber,LocationoftheElevatingDevice,ElevatingDevicesLicenseNumber,LICENSESTATUS,LICENSEEXPIRYDATE,LICENSEHOLDER,LICENSEHOLDERACCOUNTNUMBER,LICENSEHOLDERADDRESS,BILLINGCUSTOMER,BILLINGADDRESS,BILLINGACCOUNT,Elevating devices number,Owner Name,Owner Address,Owner Account Number,Device Class,Device Type,DeviceStatus,Location of Device,under review
0,10,111 WELLESLEY ST W TORONTO M7A 1A2 ON CA,EDLIC-000010,ACTIVE,28-Apr-17,LEGISLATIVE ASSEMBLY OF ONTARIO ATTN: JOHN ED...,data redacted,99 WELLESLEY ST W WHITNEY BLOCK ROOM 2540 TOR...,LEGISLATIVE ASSEMBLY OF ONTARIO ATTN: JOHN ED...,99 WELLESLEY ST W WHITNEY BLOCK ROOM 2540 TOR...,data redacted,10,LEGISLATIVE ASSEMBLY OF ONTARIO ATTN: JOHN ED...,99 WELLESLEY ST W WHITNEY BLOCK ROOM 2540 TOR...,redacted,Elevators,Passenger Elevator,Active,111 WELLESLEY ST W TORONTO M7A 1A2 ON CA,N
1,1009,404 MAIN ST WOODSTOCK N4S 7X5 ON CA,EDLIC-001009,ACTIVE,15-Jul-17,AGRIBRANDS PURINA CANADA INC,data redacted,404 MAIN ST PO BOX 250 WOODSTOCK ON N4S 7X5 CA,AGRIBRANDS PURINA CANADA INC,"404 MAIN ST PO BOX 250 WOODSTOCK, ON, N4S 7X5, CA",data redacted,1009,AGRIBRANDS PURINA CANADA INC,404 MAIN ST PO BOX 250 WOODSTOCK ON N4S 7X5 CA,redacted,Elevators,Freight Elevator,Active,404 MAIN ST WOODSTOCK N4S 7X5 ON CA,N
2,10145,45 SECOND ST E CORNWALL K6H 1V5 ON CA,EDLIC-010145,ACTIVE,15-Mar-17,CORNWALL PUBLIC LIBRARY,data redacted,45 SECOND ST E P.O.BOX 939 CORNWALL ON K6H 5V1 CA,CORNWALL PUBLIC LIBRARY,"45 SECOND ST E P.O.BOX 939 CORNWALL, ON, K6H 5...",data redacted,10145,CORNWALL PUBLIC LIBRARY,45 SECOND ST E P.O.BOX 939 CORNWALL ON K6H 5V1 CA,redacted,Elevators,Freight Elevator,Active,45 SECOND ST E CORNWALL K6H 1V5 ON CA,N
3,1018,150 SIMCOE ST LONDON N6A 4M3 ON CA,EDLIC-001018,ACTIVE,15-Oct-17,LABATT LONDON -LONDON BREWERY,data redacted,150 SIMCOE ST PO BOX 5050 STN B LONDON ON N6A ...,LABATT LONDON -LONDON BREWERY,"150 SIMCOE ST PO BOX 5050 STN B LONDON, ON, N6...",data redacted,1018,LABATT LONDON -LONDON BREWERY,150 SIMCOE ST PO BOX 5050 STN B LONDON ON N6A ...,redacted,Elevators,Passenger Elevator,Active,150 SIMCOE ST LONDON N6A 4M3 ON CA,N
4,1019,150 SIMCOE ST LONDON N6A 4M3 ON CA,EDLIC-001019,ACTIVE,15-Oct-17,LABATT LONDON -LONDON BREWERY,data redacted,150 SIMCOE ST PO BOX 5050 STN B LONDON ON N6A ...,LABATT LONDON -LONDON BREWERY,"150 SIMCOE ST PO BOX 5050 STN B LONDON, ON, N6...",data redacted,1019,LABATT LONDON -LONDON BREWERY,150 SIMCOE ST PO BOX 5050 STN B LONDON ON N6A ...,redacted,Elevators,Passenger Elevator,Active,150 SIMCOE ST LONDON N6A 4M3 ON CA,N


## AND-102 Task 5: Merge 1, License + Installed — Part 3

In [5]:
# Step 1: Compare location columns — identical vs different
identical_mask = merged_df['LocationoftheElevatingDevice'] == merged_df['Location of Device']

print(f'Identical location values: {identical_mask.sum():,}')
print(f'Different location values:  {(~identical_mask).sum():,}')
print(f'Match rate: {identical_mask.mean():.2%}')

Identical location values: 43,144
Different location values:  153
Match rate: 99.65%


In [6]:
# Step 2: Inspect rows where location values disagree
discrepancies = merged_df[~identical_mask][['ElevatingDevicesNumber',
                                               'LocationoftheElevatingDevice',
                                               'Location of Device']].reset_index(drop=True)
discrepancies.head(10)

,ElevatingDevicesNumber,LocationoftheElevatingDevice,Location of Device
0,11874,4 LATIMER AV TORONTO M5N 2L9 ON CA,4 LATIMER AVE TORONTO M5N 2L9 ON CA
1,13052,67 PARKWOODS VILLAGE DR DON MILLS M3A 2X4 ON CA,67 PARKWOODS VILLAGE DR NORTH YORK M3A 2X4 ON CA
2,13053,67 PARKWOODS VILLAGE DR DON MILLS M3A 2X4 ON CA,67 PARKWOODS VILLAGE DR NORTH YORK M3A 2X4 ON CA
3,13249,70 PARKWOODS VILLAGE DR DON MILLS M3A 2X7 ON CA,70 PARKWOODS VILLAGE DR NORTH YORK M3A 2X7 ON CA
4,15251,NaN,NaN
5,15474,NaN,NaN
6,16417,31 ST DENNIS DR DON MILLS M3C 1G7 ON CA,31 ST DENNIS DR NORTH YORK M3C 1G7 ON CA
7,16418,31 ST DENNIS DR DON MILLS M3C 1G7 ON CA,31 ST DENNIS DR NORTH YORK M3C 1G7 ON CA
8,16524,35 ST DENNIS DR DON MILLS M3C 1G9 ON CA,35 ST DENNIS DR NORTH YORK M3C 1G9 ON CA
9,16525,35 ST DENNIS DR DON MILLS M3C 1G9 ON CA,35 ST DENNIS DR NORTH YORK M3C 1G9 ON CA


In [7]:
# Step 3: Extract postal codes from each location column
# Canadian postal code format: A1A 1A1
import re

postal_pattern = r'[A-Z]\d[A-Z]\s\d[A-Z]\d'

merged_df = merged_df.copy()
merged_df['postal_license']   = merged_df['LocationoftheElevatingDevice'].str.extract(f'({postal_pattern})')
merged_df['postal_installed'] = merged_df['Location of Device'].str.extract(f'({postal_pattern})')

disc = merged_df[~identical_mask]
postal_match = disc['postal_license'] == disc['postal_installed']

print(f'Among {len(disc)} discrepant rows:')
print(f'  Postal codes match (abbreviation/city-name variant): {postal_match.sum()}')
print(f'  Postal codes differ (genuine address mismatch):       {(~postal_match).sum()}')

Among 153 discrepant rows:
  Postal codes match (abbreviation/city-name variant): 24
  Postal codes differ (genuine address mismatch):       129


In [8]:
rows_before = len(merged_df)

postal_match_all = merged_df['postal_license'] == merged_df['postal_installed']
merged_df = merged_df[postal_match_all].reset_index(drop=True)

print(f'Rows before postal filter: {rows_before:,}')
print(f'Rows after postal filter:  {len(merged_df):,}')
print(f'Rows removed:              {rows_before - len(merged_df):,}')

merged_df.head()

Rows before postal filter: 43,297
Rows after postal filter:  43,154
Rows removed:              143


,ElevatingDevicesNumber,LocationoftheElevatingDevice,ElevatingDevicesLicenseNumber,LICENSESTATUS,LICENSEEXPIRYDATE,LICENSEHOLDER,LICENSEHOLDERACCOUNTNUMBER,LICENSEHOLDERADDRESS,BILLINGCUSTOMER,BILLINGADDRESS,...,Owner Name,Owner Address,Owner Account Number,Device Class,Device Type,DeviceStatus,Location of Device,under review,postal_license,postal_installed
0,10,111 WELLESLEY ST W TORONTO M7A 1A2 ON CA,EDLIC-000010,ACTIVE,28-Apr-17,LEGISLATIVE ASSEMBLY OF ONTARIO ATTN: JOHN ED...,data redacted,99 WELLESLEY ST W WHITNEY BLOCK ROOM 2540 TOR...,LEGISLATIVE ASSEMBLY OF ONTARIO ATTN: JOHN ED...,99 WELLESLEY ST W WHITNEY BLOCK ROOM 2540 TOR...,...,LEGISLATIVE ASSEMBLY OF ONTARIO ATTN: JOHN ED...,99 WELLESLEY ST W WHITNEY BLOCK ROOM 2540 TOR...,redacted,Elevators,Passenger Elevator,Active,111 WELLESLEY ST W TORONTO M7A 1A2 ON CA,N,M7A 1A2,M7A 1A2
1,1009,404 MAIN ST WOODSTOCK N4S 7X5 ON CA,EDLIC-001009,ACTIVE,15-Jul-17,AGRIBRANDS PURINA CANADA INC,data redacted,404 MAIN ST PO BOX 250 WOODSTOCK ON N4S 7X5 CA,AGRIBRANDS PURINA CANADA INC,"404 MAIN ST PO BOX 250 WOODSTOCK, ON, N4S 7X5, CA",...,AGRIBRANDS PURINA CANADA INC,404 MAIN ST PO BOX 250 WOODSTOCK ON N4S 7X5 CA,redacted,Elevators,Freight Elevator,Active,404 MAIN ST WOODSTOCK N4S 7X5 ON CA,N,N4S 7X5,N4S 7X5
2,10145,45 SECOND ST E CORNWALL K6H 1V5 ON CA,EDLIC-010145,ACTIVE,15-Mar-17,CORNWALL PUBLIC LIBRARY,data redacted,45 SECOND ST E P.O.BOX 939 CORNWALL ON K6H 5V1 CA,CORNWALL PUBLIC LIBRARY,"45 SECOND ST E P.O.BOX 939 CORNWALL, ON, K6H 5...",...,CORNWALL PUBLIC LIBRARY,45 SECOND ST E P.O.BOX 939 CORNWALL ON K6H 5V1 CA,redacted,Elevators,Freight Elevator,Active,45 SECOND ST E CORNWALL K6H 1V5 ON CA,N,K6H 1V5,K6H 1V5
3,1018,150 SIMCOE ST LONDON N6A 4M3 ON CA,EDLIC-001018,ACTIVE,15-Oct-17,LABATT LONDON -LONDON BREWERY,data redacted,150 SIMCOE ST PO BOX 5050 STN B LONDON ON N6A ...,LABATT LONDON -LONDON BREWERY,"150 SIMCOE ST PO BOX 5050 STN B LONDON, ON, N6...",...,LABATT LONDON -LONDON BREWERY,150 SIMCOE ST PO BOX 5050 STN B LONDON ON N6A ...,redacted,Elevators,Passenger Elevator,Active,150 SIMCOE ST LONDON N6A 4M3 ON CA,N,N6A 4M3,N6A 4M3
4,1019,150 SIMCOE ST LONDON N6A 4M3 ON CA,EDLIC-001019,ACTIVE,15-Oct-17,LABATT LONDON -LONDON BREWERY,data redacted,150 SIMCOE ST PO BOX 5050 STN B LONDON ON N6A ...,LABATT LONDON -LONDON BREWERY,"150 SIMCOE ST PO BOX 5050 STN B LONDON, ON, N6...",...,LABATT LONDON -LONDON BREWERY,150 SIMCOE ST PO BOX 5050 STN B LONDON ON N6A ...,redacted,Elevators,Passenger Elevator,Active,150 SIMCOE ST LONDON N6A 4M3 ON CA,N,N6A 4M3,N6A 4M3


## AND-102 Task 5: Merge 1, License + Installed — Part 4

In [9]:
# Preliminary analysis: scan all installed columns for category counts
for col in installed_df.columns:
    n = installed_df[col].nunique(dropna=False)
    print(f'{col!r}: {n} unique values')

'Elevating devices number': 46936 unique values
'Owner Name': 16586 unique values
'Owner Address': 13339 unique values
'Owner Account Number': 1 unique values
'Device Class': 1 unique values
'Device Type': 11 unique values
'DeviceStatus': 5 unique values
'Location of Device': 24029 unique values
'under review': 2 unique values


In [10]:

print(installed_df['Device Type'].value_counts(dropna=False))

Device Type
Passenger Elevator      42405
Freight Elevator         2912
LULA Elevator            1254
Observation Elevator      325
Freight Elevator-P         17
Freight Elevator-E         11
Sidewalk Elevator           5
Temporary Elevator          4
Material Lift - ATD         1
Special Installation        1
Power Type Manlift          1
Name: count, dtype: int64


# Identify problematic categories
-Freight Elevator-P' and 'Freight Elevator-E': cryptic suffixed variants of Freight Elevator

-Material Lift - ATD', 'Special Installation', 'Power Type Manlift': single-occurrence values


In [11]:
# Clean 'Device Type':

#   - Merge 'Freight Elevator-P' and 'Freight Elevator-E' into 'Freight Elevator'
#     Reason: suffixes -P/-E are undocumented subtypes with only 28 total rows;
#     consolidating avoids misleading sparse categories in analysis.


#   - Map the three single-occurrence values to 'Specialized Equipment'
#     Reason: 1-row categories have no statistical weight and clutter groupby operations.
#   Result: 11 categories → 8 clean, meaningful categories.

remap = {
    'Freight Elevator-P': 'Freight Elevator',
    'Freight Elevator-E': 'Freight Elevator',
    'Material Lift - ATD': 'Specialized Equipment',
    'Special Installation': 'Specialized Equipment',
    'Power Type Manlift': 'Specialized Equipment',
}

merged_df['Device Type'] = merged_df['Device Type'].replace(remap)

print(f'Categories before: 11')
print(f'Categories after:  {merged_df["Device Type"].nunique()}')
print()
print(merged_df['Device Type'].value_counts(dropna=False))

Categories before: 11
Categories after:  7

Device Type
Passenger Elevator       39883
Freight Elevator          1780
LULA Elevator             1180
Observation Elevator       303
Temporary Elevator           4
Sidewalk Elevator            2
Specialized Equipment        2
Name: count, dtype: int64


In [12]:
# Merge 1 end — final shape and explanation
print(f'Merge 1 final shape: {merged_df.shape}')
print()
print('Rows lost:')
print('  - 3,639 from installed_df: devices with no active or pending license')
print('    excluded by the inner join — outside the scope of this pipeline.')
print('  - 143 from the postal code filter, broken into two causes:')
print('      129 rows where postal codes genuinely differed between the license')
print('      and installed location columns — real address mismatches.')
print('      14 rows where the regex could not extract a postal code from one')
print('      or both columns (null/malformed address). In pandas, NaN == NaN')
print('      evaluates to False, so these rows also fail the filter.')

Merge 1 final shape: (43154, 22)

Rows lost:
  - 3,639 from installed_df: devices with no active or pending license
    excluded by the inner join — outside the scope of this pipeline.
  - 143 from the postal code filter, broken into two causes:
      129 rows where postal codes genuinely differed between the license
      and installed location columns — real address mismatches.
      14 rows where the regex could not extract a postal code from one
      or both columns (null/malformed address). In pandas, NaN == NaN
      evaluates to False, so these rows also fail the filter.


## AND-102 Task 5: Merge 2, License + Installed + Altered

In [13]:
# Merge 2 start — row count entering this merge
print(f'merged_df entering Merge 2: {len(merged_df):,} rows')

merged_df entering Merge 2: 43,154 rows


In [14]:
# Load altered.json
altered_df = pd.read_json('../data/altered.json')

print(altered_df.shape)
print(altered_df.columns.tolist())
altered_df.head()

(31619, 10)
['originating service request number', 'Alteration Customer', 'Summary', 'Elevating Devices Number', 'Inspection number', 'Alteration  Location', 'Alteration Type', 'Status of Alteration Request', 'Alteration contractor name', 'Billing Customer']


,originating service request number,Alteration Customer,Summary,Elevating Devices Number,Inspection number,Alteration Location,Alteration Type,Status of Alteration Request,Alteration contractor name,Billing Customer
0,782682,LEGISLATIVE ASSEMBLY OF ONTARIO ATTN: JOHN ED...,Minor B Alteration Machine Guarding,8,NaN,111 WELLESLEY ST W TORONTO M7A 1A2 ON CA,ED-Minor B Alteration,Passed,NaN,THYSSENKRUPP ELEVATOR (CANADA) LTD
1,834770,LEGISLATIVE ASSEMBLY OF ONTARIO ATTN: JOHN ED...,"Car top Railing, 12-2191A/05",8,NaN,111 WELLESLEY ST W TORONTO M7A 1A2 ON CA,ED-Minor A Alteration,Passed,NaN,THYSSENKRUPP ELEVATOR (CANADA) LTD
2,1357738,LEGISLATIVE ASSEMBLY OF ONTARIO ATTN: JOHN ED...,Major Alteration 6028233 Car 4,8,NaN,111 WELLESLEY ST W TORONTO M7A 1A2 ON CA,ED-Major Alteration,Passed,NaN,KONE INC - TORONTO SEB
3,834774,LEGISLATIVE ASSEMBLY OF ONTARIO ATTN: JOHN ED...,"Car top Railing, 12-2191A/05",9,NaN,111 WELLESLEY ST W TORONTO M7A 1A2 ON CA,ED-Minor A Alteration,Passed,NaN,THYSSENKRUPP ELEVATOR (CANADA) LTD
4,782688,LEGISLATIVE ASSEMBLY OF ONTARIO ATTN: JOHN ED...,Minor B Alteration Machine Guarding,9,NaN,111 WELLESLEY ST W TORONTO M7A 1A2 ON CA,ED-Minor B Alteration,Passed,NaN,THYSSENKRUPP ELEVATOR (CANADA) LTD


In [15]:
rows_before = len(merged_df)

merged_df = merged_df.merge(
    altered_df,
    left_on='ElevatingDevicesNumber',
    right_on='Elevating Devices Number',
    how='left'
)

print(f'Rows before merge: {rows_before:,}')
print(f'Rows after merge:  {len(merged_df):,}')
print(f'Rows gained:       {len(merged_df) - rows_before:,}')
print(f'Elevators with no alteration record: {merged_df["Elevating Devices Number"].isna().sum():,}')

Rows before merge: 43,154
Rows after merge:  52,339
Rows gained:       9,185
Elevators with no alteration record: 21,158


In [16]:
# Elevators with 5 or more alterations
alteration_counts = (
    merged_df.dropna(subset=['Elevating Devices Number'])
    .groupby('ElevatingDevicesNumber')
    .size()
    .reset_index(name='alteration_count')
)

heavy_alts = alteration_counts[alteration_counts['alteration_count'] >= 5]
fleet_size = merged_df['ElevatingDevicesNumber'].nunique()

print(f'Elevators with 5+ alterations: {len(heavy_alts):,}')
print(f'Fleet size (unique elevators):  {fleet_size:,}')
print(f'Proportion of fleet:            {len(heavy_alts) / fleet_size:.2%}')
print()
print('Top 10 most altered elevators:')
print(alteration_counts.sort_values('alteration_count', ascending=False).head(10).to_string(index=False))

Elevators with 5+ alterations: 51
Fleet size (unique elevators):  43,154
Proportion of fleet:            0.12%

Top 10 most altered elevators:
 ElevatingDevicesNumber  alteration_count
                  14191                 7
                  61456                 7
                  28465                 6
                  23693                 6
                  32113                 6
                  32112                 6
                  35118                 6
                  17922                 6
                  23692                 6
                  23209                 6


In [17]:
# Merge 2 end — final shape and explanation
print(f'Merge 2 final shape: {merged_df.shape}')
print()
print('Rows gained:')
print('  - altered.json has one row per alteration event.')
print('    Elevators altered multiple times generate multiple rows in the join.')
print('  - Elevators never altered are kept with NaN in alteration columns (left join).')

Merge 2 final shape: (52339, 32)

Rows gained:
  - altered.json has one row per alteration event.
    Elevators altered multiple times generate multiple rows in the join.
  - Elevators never altered are kept with NaN in alteration columns (left join).


## AND-102 Task 5: Merge 3, License + Installed + Altered + Inspections — Part 1

In [18]:
# Merge 3 start — row count entering this merge
print(f'merged_df entering Merge 3: {len(merged_df):,} rows')

merged_df entering Merge 3: 52,339 rows


In [19]:
inspection_df = pd.read_csv('../data/inspection.csv')

print(inspection_df.shape)
print(inspection_df.columns.tolist())
inspection_df.head()

(143181, 9)
['originatingservicerequestnumber', 'InspectionCustomer', 'ElevatingDevicesNumber', 'InspectionNumber', 'InspectionLocation', 'InspectionType', 'Earliest_INSPECTION_Date', 'Latest_INSPECTION_Date', 'InspectionOutcome']


,originatingservicerequestnumber,InspectionCustomer,ElevatingDevicesNumber,InspectionNumber,InspectionLocation,InspectionType,Earliest_INSPECTION_Date,Latest_INSPECTION_Date,InspectionOutcome
0,55620,LANDSCAPE COURT APARTMENTS LTD,9948,3157237,10 WYCOMBE RD DOWNSVIEW M3M 2W7 ON CA,ED-Followup Inspection,1/10/2011,1/10/2011,Passed
1,357196,YORK CONDO CORP #255,28588,2823870,75 EMMETT AV TORONTO M6M 5A7 ON CA,ED-Periodic Inspection,1/10/2011,1/10/2011,Follow up
2,102922,BMCC REIT,76833,3158421,5099 CREEKBANK RD MISSISSAUGA L4W 1R3 ON CA,ED-Followup Inspection,1/10/2011,1/10/2011,Passed
3,357352,YORK CONDO CORP #255,28590,2823948,75 EMMETT AV TORONTO M6M 5A7 ON CA,ED-Periodic Inspection,1/10/2011,1/10/2011,Follow up
4,516054,YORK CONDO CORP #255,28589,3193568,75 EMMETT AV TORONTO M6M 5A7 ON CA,ED-Periodic Inspection,1/10/2011,1/10/2011,Follow up


In [20]:
# Can one elevator have multiple inspections?
per_elevator = inspection_df.groupby('ElevatingDevicesNumber')['InspectionNumber'].count()

print('Inspections per elevator:')
print(f'  Min:  {per_elevator.min()}')
print(f'  Max:  {per_elevator.max()}')
print(f'  Mean: {per_elevator.mean():.2f}')
print(f'  Elevators with >1 inspection: {(per_elevator > 1).sum():,} of {len(per_elevator):,}')
print()

# Can one inspection cover multiple elevators?
per_inspection = inspection_df.groupby('InspectionNumber')['ElevatingDevicesNumber'].count()

print('Elevators per inspection:')
print(f'  Min:  {per_inspection.min()}')
print(f'  Max:  {per_inspection.max()}')
print(f'  Mean: {per_inspection.mean():.2f}')
print(f'  Inspections covering >1 elevator: {(per_inspection > 1).sum():,} of {len(per_inspection):,}')

Inspections per elevator:
  Min:  1
  Max:  24
  Mean: 3.50
  Elevators with >1 inspection: 33,480 of 40,954

Elevators per inspection:
  Min:  1
  Max:  1
  Mean: 1.00
  Inspections covering >1 elevator: 0 of 143,181


**Findings:**

One elevator can absolutely have multiple inspections — on average 3.5 per elevator, with some reaching as many as 24. That makes sense: inspections happen periodically over the life of a device, so repeat records are expected, not errors.

The other direction is clean: every InspectionNumber maps to exactly one elevator. No inspection covers more than one device. The relationship is strictly one-to-many (elevator → inspections).

This means the merge will expand the row count, one row per inspection event per elevator. Elevators with no inspection record will keep a single row with NaN in the inspection columns.

## AND-102 Task 5: Merge 3, License + Installed + Altered + Inspections — Part 2

**Approach: aggregate inspection history per elevator**

A straight left join would expand rows to ~180K — one per inspection event — which breaks the one-row-per-elevator structure needed for the dashboard. Keeping only the most recent inspection is simple but throws away history.

Aggregating before the join gives the best of both: one row per elevator, but richer data. For each elevator we compute total inspections, date range, last outcome, and pass rate. Pass rate is defined as the proportion of outcomes that are unambiguously positive: `Passed`, `Passed Major`, `Passed Sub`, `All Orders Resolved`, `Complete`, `Inspection Complete`, or `Complete Enforcement`.

In [21]:
inspection_df['Latest_INSPECTION_Date'] = pd.to_datetime(
    inspection_df['Latest_INSPECTION_Date'], format='%m/%d/%Y'
)
inspection_df['Earliest_INSPECTION_Date'] = pd.to_datetime(
    inspection_df['Earliest_INSPECTION_Date'], format='%m/%d/%Y'
)

PASS_OUTCOMES = {
    'Passed', 'Passed Major', 'Passed Sub',
    'All Orders Resolved', 'Complete',
    'Inspection Complete', 'Complete Enforcement'
}

inspection_df['is_passed'] = inspection_df['InspectionOutcome'].isin(PASS_OUTCOMES)

# Most recent outcome per elevator (for last_outcome column)
last_outcome = (
    inspection_df
    .sort_values('Latest_INSPECTION_Date', ascending=False)
    .drop_duplicates(subset='ElevatingDevicesNumber', keep='first')
    [['ElevatingDevicesNumber', 'InspectionOutcome', 'InspectionType']]
    .rename(columns={
        'InspectionOutcome': 'last_inspection_outcome',
        'InspectionType': 'last_inspection_type'
    })
)

# Aggregate metrics per elevator
inspection_agg = (
    inspection_df
    .groupby('ElevatingDevicesNumber')
    .agg(
        total_inspections=('InspectionNumber', 'count'),
        first_inspection_date=('Earliest_INSPECTION_Date', 'min'),
        last_inspection_date=('Latest_INSPECTION_Date', 'max'),
        pct_passed=('is_passed', 'mean')
    )
    .reset_index()
)

inspection_agg = inspection_agg.merge(last_outcome, on='ElevatingDevicesNumber', how='left')
inspection_agg['pct_passed'] = inspection_agg['pct_passed'].round(2)

print(inspection_agg.shape)
inspection_agg.head()

(40954, 7)


,ElevatingDevicesNumber,total_inspections,first_inspection_date,last_inspection_date,pct_passed,last_inspection_outcome,last_inspection_type
0,8,6,2012-03-05,2015-03-27,0.17,Follow up,ED-Periodic Inspection
1,9,7,2012-03-05,2015-03-27,0.14,DC Follow up,ED-Periodic Inspection
2,10,9,2012-03-05,2015-03-27,0.33,All Orders Resolved,ED-Sub Inspection Major
3,11,10,2012-03-05,2015-03-27,0.30,All Orders Resolved,ED-Sub Inspection Major
4,13,8,2012-03-06,2015-04-30,0.25,All Orders Resolved,ED-Followup Inspection


In [22]:
rows_before = len(merged_df)

merged_df = merged_df.merge(
    inspection_agg,
    on='ElevatingDevicesNumber',
    how='left'
)

print(f'Rows before merge: {rows_before:,}')
print(f'Rows after merge:  {len(merged_df):,}')
print(f'Rows gained/lost:  {len(merged_df) - rows_before:,}')
print(f'Elevators with no inspection record: {merged_df["total_inspections"].isna().sum():,}')

Rows before merge: 52,339
Rows after merge:  52,339
Rows gained/lost:  0
Elevators with no inspection record: 3,553


In [23]:
# Merge 3 end — final shape and explanation
print(f'Merge 3 final shape: {merged_df.shape}')
print()
print('Rows gained/lost: 0')
print('  - inspection_df (143,181 rows) was aggregated to one row per elevator')
print('    before the join, so the left join does not expand rows.')
print('  - Elevators with no inspection history are kept with NaN (left join).')

Merge 3 final shape: (52339, 38)

Rows gained/lost: 0
  - inspection_df (143,181 rows) was aggregated to one row per elevator
    before the join, so the left join does not expand rows.
  - Elevators with no inspection history are kept with NaN (left join).


## AND-102 Task 5: Merge 3, License + Installed + Altered + Inspections — Part 3

Three columns are dropped before saving:

- **** and ****: both are duplicates of , left behind as residual join keys from Merge 1 and Merge 2 respectively.
- **** and ****: temporary columns created only for the location filter in Part 3. They have no analytical value in the final dataset.

In [24]:
cols_to_drop = [
    "Elevating devices number",
    "Elevating Devices Number",
    "postal_license",
    "postal_installed",
]

merged_df = merged_df.drop(columns=cols_to_drop)

print(f"Columns dropped: {cols_to_drop}")
print(f"Final shape: {merged_df.shape}")
merged_df.head()

Columns dropped: ['Elevating devices number', 'Elevating Devices Number', 'postal_license', 'postal_installed']
Final shape: (52339, 34)


,ElevatingDevicesNumber,LocationoftheElevatingDevice,ElevatingDevicesLicenseNumber,LICENSESTATUS,LICENSEEXPIRYDATE,LICENSEHOLDER,LICENSEHOLDERACCOUNTNUMBER,LICENSEHOLDERADDRESS,BILLINGCUSTOMER,BILLINGADDRESS,...,Alteration Type,Status of Alteration Request,Alteration contractor name,Billing Customer,total_inspections,first_inspection_date,last_inspection_date,pct_passed,last_inspection_outcome,last_inspection_type
0,10,111 WELLESLEY ST W TORONTO M7A 1A2 ON CA,EDLIC-000010,ACTIVE,28-Apr-17,LEGISLATIVE ASSEMBLY OF ONTARIO ATTN: JOHN ED...,data redacted,99 WELLESLEY ST W WHITNEY BLOCK ROOM 2540 TOR...,LEGISLATIVE ASSEMBLY OF ONTARIO ATTN: JOHN ED...,99 WELLESLEY ST W WHITNEY BLOCK ROOM 2540 TOR...,...,ED-Minor B Alteration,Passed,NaN,THYSSENKRUPP ELEVATOR (CANADA) LTD,9.0,2012-03-05,2015-03-27,0.33,All Orders Resolved,ED-Sub Inspection Major
1,10,111 WELLESLEY ST W TORONTO M7A 1A2 ON CA,EDLIC-000010,ACTIVE,28-Apr-17,LEGISLATIVE ASSEMBLY OF ONTARIO ATTN: JOHN ED...,data redacted,99 WELLESLEY ST W WHITNEY BLOCK ROOM 2540 TOR...,LEGISLATIVE ASSEMBLY OF ONTARIO ATTN: JOHN ED...,99 WELLESLEY ST W WHITNEY BLOCK ROOM 2540 TOR...,...,ED-Minor A Alteration,Passed,NaN,THYSSENKRUPP ELEVATOR (CANADA) LTD,9.0,2012-03-05,2015-03-27,0.33,All Orders Resolved,ED-Sub Inspection Major
2,10,111 WELLESLEY ST W TORONTO M7A 1A2 ON CA,EDLIC-000010,ACTIVE,28-Apr-17,LEGISLATIVE ASSEMBLY OF ONTARIO ATTN: JOHN ED...,data redacted,99 WELLESLEY ST W WHITNEY BLOCK ROOM 2540 TOR...,LEGISLATIVE ASSEMBLY OF ONTARIO ATTN: JOHN ED...,99 WELLESLEY ST W WHITNEY BLOCK ROOM 2540 TOR...,...,ED-Major Alteration,Passed,NaN,THYSSENKRUPP ELEVATOR (CANADA) LTD,9.0,2012-03-05,2015-03-27,0.33,All Orders Resolved,ED-Sub Inspection Major
3,1009,404 MAIN ST WOODSTOCK N4S 7X5 ON CA,EDLIC-001009,ACTIVE,15-Jul-17,AGRIBRANDS PURINA CANADA INC,data redacted,404 MAIN ST PO BOX 250 WOODSTOCK ON N4S 7X5 CA,AGRIBRANDS PURINA CANADA INC,"404 MAIN ST PO BOX 250 WOODSTOCK, ON, N4S 7X5, CA",...,NaN,NaN,NaN,NaN,2.0,2012-01-23,2013-03-27,0.50,Passed,ED-Followup Inspection
4,10145,45 SECOND ST E CORNWALL K6H 1V5 ON CA,EDLIC-010145,ACTIVE,15-Mar-17,CORNWALL PUBLIC LIBRARY,data redacted,45 SECOND ST E P.O.BOX 939 CORNWALL ON K6H 5V1 CA,CORNWALL PUBLIC LIBRARY,"45 SECOND ST E P.O.BOX 939 CORNWALL, ON, K6H 5...",...,NaN,NaN,NaN,NaN,1.0,2016-04-27,2016-04-27,1.00,All Orders Resolved,ED-Followup Inspection


In [25]:
merged_df.to_csv('../data/merged_elevator_data.csv', index=False)

print(f'Saved to data/merged_elevator_data.csv')
print(f'Final shape: {merged_df.shape}')
print(f'Columns: {merged_df.columns.tolist()}')

Saved to data/merged_elevator_data.csv
Final shape: (52339, 34)
Columns: ['ElevatingDevicesNumber', 'LocationoftheElevatingDevice', 'ElevatingDevicesLicenseNumber', 'LICENSESTATUS', 'LICENSEEXPIRYDATE', 'LICENSEHOLDER', 'LICENSEHOLDERACCOUNTNUMBER', 'LICENSEHOLDERADDRESS', 'BILLINGCUSTOMER', 'BILLINGADDRESS', 'BILLINGACCOUNT', 'Owner Name', 'Owner Address', 'Owner Account Number', 'Device Class', 'Device Type', 'DeviceStatus', 'Location of Device', 'under review', 'originating service request number', 'Alteration Customer', 'Summary', 'Inspection number', 'Alteration  Location', 'Alteration Type', 'Status of Alteration Request', 'Alteration contractor name', 'Billing Customer', 'total_inspections', 'first_inspection_date', 'last_inspection_date', 'pct_passed', 'last_inspection_outcome', 'last_inspection_type']


---
## ETL Pipeline Summary

| Step | Operation | Rows | Columns | Key Decision |
|---|---|---|---|---|
| License load | Filter ACTIVE + PENDING_RENEWAL | 43,297 | 11 | Dropped 2,086 inactive/expired licenses |
| Installed load | — | 46,936 | 9 | — |
| **Merge 1a** | License ⋈ Installed (inner, device number) | 43,297 | 20 | 3,639 installed devices with no active license dropped |
| **Merge 1b** | Postal code location filter | 43,154 | 22 | 143 rows dropped: 129 genuine postal code mismatches + 14 where regex returned NaN |
| **Merge 2** | + Altered (left, device number) | 52,339 | 32 | Rows expanded — one row per alteration event; 21,158 elevators never altered kept with NaN |
| **Merge 3** | + Inspections (left, aggregated) | 52,339 | 34 | Row count unchanged — inspection history aggregated to one row per elevator before joining; 3,553 elevators with no inspection kept with NaN |
